Mage/src/main/java/mage/players/PlayerImpl.java

```diff
@@ -4779,10 +4779,7 @@ public boolean moveCards(Set<? extends Card> cards, Zone toZone, Ability source,
                         }
                     }
                 }
-                // TODO: must be replaced by game.getState().processAction(game), see isInUseableZoneDiesTrigger comments
```

Comment:
Must keep todo, it’s still actual. Multi-steps effect’s must call process action between steps to process state base effects/triggers/control/etc (example: “move xxx to battlefield, then do something else”):
* Multi-steps by multiple effects - ability(effect1, effect2): already call it between apply effects on resolve;
* Multi-steps by one big effect - ability(effect):
  * hidden use after move to battlefield (todo)
  * manually use after moving to other zones;

Reply:
I don't understand what you are referring to.

I removed the TODO here because the isInUseableZoneDiesTrigger / short living LKI problem has been addressed, so leaving that comment would be confusing.

1. Do you think that this method should be calling `game.getState().processAction(game)` instead of `game.applyEffects()`? If so, why? (see #12447 )
2. Do you think this method should not be calling `game.applyEffects()` at all? If so, where should it be getting handled instead? (see #12448 )
3. Do you agree or disagree that the logic should be the same for both cards and tokens?

Either way, if you think something needs to be different still, make new issue to explain in detail please.

Reply:
Cause `isInUseableZoneDiesTrigger` contains detailed description of the problems (including short living lki) and why it must be fixed and how.

Yes, it must call processAction after move to battlefield cause ability's resolve code call it between effects. Devs must use same things in multi steps effect but miss it. Alternative solution: find all move to battlefield usages and call processAction  manually.

![shot_240610_062132](https://github.com/magefree/mage/assets/8344157/3b5440ba-4b5e-4d8c-829b-15880efe928d)

Reply:
I still don't understand what you think the problem is.

Reply:
The problem in uncontrolled use of magic "apply effects" methods that was able to "fix any problem" (and it's literally fix it). As a result -- it's all around the code and can't be refactored/fixed/removed without break a dozen things (things that was wrongly written from the start).

What must be called:
![shot_240610_073901](https://github.com/magefree/mage/assets/8344157/013a0c44-b3c8-426b-858d-63f54c599c4e)

Reply:
Flicker implementation example:
> Exile target nontoken permanent, then return it to the battlefield under its owner's control.

By two effects:
![shot_240610_134446](https://github.com/magefree/mage/assets/8344157/a09ea9b7-35bc-4416-befe-2af1b77d0e93)

By single effect with two multi-steps:
![shot_240610_135753](https://github.com/magefree/mage/assets/8344157/03a0191c-be8e-4376-85df-d01bfc160bfd)

Reply:
BTW current flicker implementation looks bugged cause it move cards one by one instead all at once

Must be like that:
![shot_240610_135941](https://github.com/magefree/mage/assets/8344157/25c1688a-1392-45e7-9fa4-bfb29939510d)

Reply:
Ideally must check new events after each step, but it require too much code:
![shot_240610_141837](https://github.com/magefree/mage/assets/8344157/fc80f29a-397b-442b-8385-b0794070d493)

Reply:
BTW that's Bitterheart Witch is bugged too - miss check for attach legality:

> When Bitterheart Witch dies, you may search your library for a Curse card, put it onto the battlefield attached to target player, then shuffle.

> The Curse must be legally able to enchant the player. For example, if the player has protection from red, you couldn’t put a red Curse onto the battlefield this way. (2011-09-22)

Reply:
And of course, if you start to use ApplyEffects everywhere, then the event/triggers will see the wrong state of the permanents (not at the time of the event, but at a the random moment after the last call to ApplyEffect). That’s why it’s important to process/check all events/triggers as fast as possible by processAction.

Reply:
> The problem in uncontrolled use of magic "apply effects" methods that was able to "fix any problem" (and it's literally fix it). As a result -- it's all around the code and can't be refactored/fixed/removed without break a dozen things (things that was wrongly written from the start).

Seems like a matter of opinion. Clearly you think it's the wrong paradigm, but I don't see a compelling justification for totally reworking the engine.
But as I stated, only remaining use of `game.applyEffects()` is in inner engine code. Do you think all of those are bad approach? Should `applyEffects()` _ever_ be called on its own without processAction, or is every single instance a workaround?
I agree that there are some rare bugs that need processAction between steps, but I figure those can be solved as reported. Yet you object to adding processAction (e.g. Rain of Riches) even when it is clearly necessary to comply with mtg rules, so what makes it bad in some cases and good in others?
Good style for devs - must be handled correctly in inner logic as much as possible, so only weird complicated custom effects need to call it manually.

> Flicker implementation example:

The code in your screenshot doesn't match master, that was refactored to a single effect back in #10564

> BTW current flicker implementation looks bugged cause it move cards one by one instead all at once

I think it's the way it is because it's possible for the cards to have different owners and thus come back under different player's control. Maybe can be improved with some research. But is it a real bug? Need a test case that gets handled wrongly.

> And of course, if you start to use ApplyEffects everywhere, then the event/triggers will see the wrong state of the permanents (not at the time of the event, but at a the random moment after the last call to ApplyEffect). That’s why it’s important to process/check all events/triggers as fast as possible by processAction.

Well, there are only 25 usages here. You can try to investigate which should be replaced with processAction or which should be removed entirely.

``` Abstract method
    applyEffects()
Usages or usages of base method in All Places  (25 usages found)
    Unclassified  (25 usages found)
        mage (1)  (17 usages found)
            mage.abilities  (1 usage found)
                AbilityImpl  (1 usage found)
                    activate(Game, boolean)  (1 usage found)
                        262 game.applyEffects();
            mage.abilities.effects  (2 usages found)
                AuraReplacementEffect  (2 usages found)
                    replaceEvent(GameEvent, Ability, Game)  (2 usages found)
                        82 game.applyEffects(); // So continuousEffects are removed if previous effect of the same ability did move objects that cause continuous effects
                        152 game.applyEffects();
            mage.game  (9 usages found)
                GameImpl  (6 usages found)
                    playPriority(UUID, boolean)  (4 usages found)
                        1676 applyEffects();
                        1697 applyEffects();
                        1711 applyEffects();
                        1729 applyEffects();
                    leave(UUID)  (1 usage found)
                        3347 applyEffects(); // to remove control from effects removed meanwhile
                    cheat(UUID, List<Card>, List<Card>, List<PutToBattlefieldInfo>, List<Card>, List<Card>, List<Card>)  (1 usage found)
                        3759 applyEffects();
                GameState  (3 usages found)
                    processAction(Game)  (1 usage found)
                        681 game.applyEffects();
                    removeEocEffects(Game)  (1 usage found)
                        701 game.applyEffects();
                    removeEotEffects(Game)  (1 usage found)
                        709 game.applyEffects();
            mage.game.permanent  (1 usage found)
                PermanentImpl  (1 usage found)
                    transform(Ability, Game, boolean)  (1 usage found)
                        694 game.applyEffects(); // not process action - no firing of simultaneous events yet
            mage.game.permanent.token  (1 usage found)
                TokenImpl  (1 usage found)
                    putOntoBattlefieldHelper(CreateTokenEvent, Game, Ability, boolean, boolean, UUID, UUID, ...)  (1 usage found)
                        457 game.applyEffects(); // without LKI reset
            mage.game.turn  (2 usages found)
                DrawStep  (1 usage found)
                    beginStep(Game, UUID)  (1 usage found)
                        31 game.applyEffects();
                UntapStep  (1 usage found)
                    beginStep(Game, UUID)  (1 usage found)
                        46 game.applyEffects();
            mage.players  (1 usage found)
                PlayerImpl  (1 usage found)
                    moveCards(Set<? extends Card>, Zone, Ability, Game, boolean, boolean, boolean, ...)  (1 usage found)
                        4782 game.applyEffects(); // without LKI reset
        mage-common  (1 usage found)
            mage.utils  (1 usage found)
                SystemUtil  (1 usage found)
                    putCardToZone(Ability, Game, Player, Card, Zone)  (1 usage found)
                        822 game.applyEffects();
        mage-player-ai-ma  (7 usages found)
            mage.player.ai  (4 usages found)
                ComputerPlayer6  (2 usages found)
                    resolve(SimulationNode2, int, Game)  (1 usage found)
                        424 game.applyEffects();
                    simulatePriority(SimulationNode2, Game, int, int, int)  (1 usage found)
                        511 sim.applyEffects();
                SimulatedPlayer2  (2 usages found)
                    triggerAbility(TriggeredAbility, Game)  (1 usage found)
                        315 game.applyEffects();
                    addAbilityNode(SimulationNode2, Ability, int, Game)  (1 usage found)
                        337 sim.applyEffects();
            mage.player.ai.util  (3 usages found)
                CombatUtil  (3 usages found)
                    willItSurvive(Game, UUID, UUID, Permanent, Permanent)  (2 usages found)
                        222 sim.applyEffects();
                        233 sim.applyEffects();
                    simulateStep(Game, Step)  (1 usage found)
                        247 game.applyEffects();
```

Reply:
> Should applyEffects() ever be called on its own without processAction, or is every single instance a workaround?

applyEffects should be called inside game engine or card effects to update game objects to actual state after previous state done. In ideally world it must have runtime check like “if game has non-processed events then raise error” to catch all wrong usages.

Reply:
I don't really understand what's the argument about here.

If there are multiple 'steps' in a OneShotEffect, there must be a call to `processAction` to then call apply, alongside handling of events, which cause updating watchers and check triggers for once the OneShotEffect is done.

![image](https://github.com/magefree/mage/assets/34709007/0f20a761-b8d6-4d41-ace5-8376489d4ee9)

The current way that is done in-between `OneShotEffect` automatically for simple cards that do have all effects split by steps in their Ability:

![image](https://github.com/magefree/mage/assets/34709007/8f4dcd68-b81b-424f-93f8-fad5e2f35b28)

But sometimes, `OneShotEffect` are created for multiple steps. They call sub-effect's `apply` by themselves, or handle multiple steps through direct player/permanent/card action methods, and it is up to the dev to call `processAction` in-between to make sure all is done step-like.

`applyEffect` should not be called directly. Or am I missing some use case?

JayDi, if your argument is that we need a big overhaul of mostly everything to no longer need to direct call processAction, we are so far from that, I don't see the point of caring about comment for an hypothetical that will certaintly never be.

Reply:
Added devs instruction with simple rules on usage of `ApplyEffects` and `ProcessAction`: #12459